### PR TITLE
[PWN-3465] Refactor/pwn 3465 solana programs

### PR DESCRIPTION
**Pull request by bigearsenal**

### CODE REVIEW

File: Sources/SolanaSwift/Programs/TokenProgram.swift [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    
+    func closeAccountInstruction(
+        account: PublicKey,
+        destination: PublicKey,
+        owner: PublicKey,
+        signers: [PublicKey]
+    ) -> TransactionInstruction
+}
+
+public struct TokenProgramImpl: SolanaBasicProgram, TokenProgram {
```
Comment (sidorov-panda) [inline Sources/SolanaSwift/Programs/TokenProgram.swift:71]: Can we rename it somehow? `TokenProgramImpl` looks not really nice

File: Sources/SolanaSwift/Programs/TokenProgram.swift [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                decimals
+            ]
+        )
+    }
+    
+    public func approveInstruction(
+        account: PublicKey,
+        delegate: PublicKey,
+        owner: PublicKey,
+        multiSigners: [Account],
```
Comment (sidorov-panda) [inline Sources/SolanaSwift/Programs/TokenProgram.swift:208]: seems like here we can pass multiSigners: [PublicKey], not [Account]

**Review by sidorov-panda — CHANGES_REQUESTED** [review]

File: Sources/SolanaSwift/Deprecated/SolanaSDK+Actions/SolanaSDK+Send.swift [hunk 3]
```diff
@@ -52,7 +52,7 @@ extension SolanaSDK {
                 .flatMap { [weak self] in
                     guard let self = self else {return .error(SolanaError.unknown)}
                     // form instruction
-                    let instruction = SystemProgram.transferInstruction(
+                    let instruction = SystemProgram().transferInstruction(
```
Comment (TrGiLong) [inline Sources/SolanaSwift/Deprecated/SolanaSDK+Actions/SolanaSDK+Send.swift:55]: Why do we initialise SystemProgram? Shouldn't it be static function?

File: Sources/SolanaSwift/Programs/TokenProgram.swift (hunk 1 above)
Comment (TrGiLong) [reply Sources/SolanaSwift/Programs/TokenProgram.swift:71]: DefaultTokenProgram is also nice

**Review by TrGiLong — COMMENTED** [review]
It would be nice if we can also add some docs to function and classes

File: Sources/SolanaSwift/Deprecated/SolanaSDK+Actions/SolanaSDK+Send.swift (hunk 3 above)
Comment (bigearsenal) [reply Sources/SolanaSwift/Deprecated/SolanaSDK+Actions/SolanaSDK+Send.swift:55]: I have done this before, but static functions may be hard to be tested. We can't make multiple instance to test function separately

Comment (TrGiLong) [reply Sources/SolanaSwift/Deprecated/SolanaSDK+Actions/SolanaSDK+Send.swift:55]: We can use [static function in protocol](https://medium.com/openclassrooms-product-design-and-engineering/injecting-and-mocking-static-frameworks-in-swift-b4fc410ab3ae).

Or using [mock library](https://github.com/Brightify/Cuckoo).

Programs are unique entity and represent an hard-coded template. It's doesn't make sense to change ProgramID or something else in transfer instruction. There should be only one single SystemProgram.

I think it maybe a special case where we don't need protocol here.

Comment (bigearsenal) [reply Sources/SolanaSwift/Deprecated/SolanaSDK+Actions/SolanaSDK+Send.swift:55]: Ok. So I will omit the protocols. Thanks (y)

**Comment by bigearsenal** [comment]
I have fixed the program to be static.
Because this one blocks another task, I will merge it right away in to 3297, we will return to it later if you have any concerns

**Pull request merged**